proto/merge.go

```
@@ -0,0 +1,114 @@
+package proto
+
+import (
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+	"golang.org/x/exp/slices"
+	protov2 "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protodesc"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/testing/protocmp"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+// MergedFileDescriptors returns a single FileDescriptorSet containing all the
+// file descriptors registered with both gogoproto and protoregistry. When
+// merging it also performs the following checks:
+// - check that all file descriptors' import paths are correct (i.e. match
+// their fully-qualified package name).
+// - check that if both gogo and protoregistry import the same file descriptor,
+// that these are identical under proto.Equal.
+func MergedFileDescriptors() (*descriptorpb.FileDescriptorSet, error) {
+	fds := &descriptorpb.FileDescriptorSet{}
+
+	// load gogo proto file descriptors
+	gogoFds := AllFileDescriptors()
+	gogoFdsMap := map[string]*descriptorpb.FileDescriptorProto{}
+	for _, compressedBz := range gogoFds {
+		rdr, err := gzip.NewReader(bytes.NewReader(compressedBz))
+		if err != nil {
+			return nil, err
+		}
+
+		bz, err := io.ReadAll(rdr)
+		if err != nil {
+			return nil, err
+		}
+
+		fd := &descriptorpb.FileDescriptorProto{}
+		err = protov2.Unmarshal(bz, fd)
+		if err != nil {
+			return nil, err
+		}
+
+		err = CheckImportPath(*fd.Name, *fd.Package)
+		if err != nil {
+			panic(err)
```

Let's return the error